pkg/api/payer/selectors/closest.go

```diff
@@ -0,0 +1,187 @@
+package selectors
+
+import (
+	"errors"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/registry"
+	"github.com/xmtp/xmtpd/pkg/topic"
+)
+
+// ClosestNodeSelectorAlgorithm selects the node with the lowest measured TCP connect latency,
+// excluding any nodes present in the banlist. Latency measurements are cached for a configured duration
+// and refreshed periodically to reduce overhead.
+type ClosestNodeSelectorAlgorithm struct {
+	reg            registry.NodeRegistry
+	preferredNodes []uint32
+	latencyCache   map[uint32]time.Duration
+	cacheMutex     sync.RWMutex
+	cacheExpiry    time.Duration
+	lastUpdate     time.Time
+	connectTimeout time.Duration
+}
+
+var _ NodeSelectorAlgorithm = (*ClosestNodeSelectorAlgorithm)(nil)
+
+func NewClosestNodeSelectorAlgorithm(
+	reg registry.NodeRegistry,
+	cacheExpiry time.Duration,
+	connectTimeout time.Duration,
+	preferredNodes ...[]uint32,
+) *ClosestNodeSelectorAlgorithm {
+	if cacheExpiry == 0 {
+		cacheExpiry = 5 * time.Minute
+	}
+	if connectTimeout == 0 {
+		connectTimeout = 2 * time.Second
+	}
+
+	var nodes []uint32
+	if len(preferredNodes) > 0 && len(preferredNodes[0]) > 0 {
+		nodes = preferredNodes[0]
+	}
+
+	return &ClosestNodeSelectorAlgorithm{
+		reg:            reg,
+		preferredNodes: nodes,
+		latencyCache:   make(map[uint32]time.Duration),
+		cacheExpiry:    cacheExpiry,
+		connectTimeout: connectTimeout,
+	}
+}
+
+func (c *ClosestNodeSelectorAlgorithm) GetNode(
+	_ topic.Topic,
+	banlist ...[]uint32,
+) (uint32, error) {
+	nodes, err := c.reg.GetNodes()
+	if err != nil {
+		return 0, err
+	}
+
+	if len(nodes) == 0 {
+		return 0, errors.New("no available nodes")
+	}
+
+	banned := make(map[uint32]struct{})
+	for _, list := range banlist {
+		for _, id := range list {
+			banned[id] = struct{}{}
+		}
+	}
+
+	// Filter nodes to preferred list if specified
+	nodesToConsider := nodes
+	if len(c.preferredNodes) > 0 {
+		preferredSet := make(map[uint32]struct{})
+		for _, nodeID := range c.preferredNodes {
+			preferredSet[nodeID] = struct{}{}
+		}
+
+		filtered := make([]registry.Node, 0, len(nodes))
+		for _, node := range nodes {
+			if _, isPreferred := preferredSet[node.NodeID]; isPreferred {
+				filtered = append(filtered, node)
+			}
+		}
+
+		// If we have preferred nodes available, use only those
+		// Otherwise fall back to all nodes
+		if len(filtered) > 0 {
+			nodesToConsider = filtered
+		}
+	}
+
+	c.cacheMutex.RLock()
+	cacheExpired := time.Since(c.lastUpdate) > c.cacheExpiry
+	c.cacheMutex.RUnlock()
+
+	if cacheExpired {
+		c.updateLatencyCache(nodesToConsider)
+	}
```

Comment:
refreshing on the hot path might cause unpredictable delays when a client thread pulls the short straw and has to refresh all nodes.